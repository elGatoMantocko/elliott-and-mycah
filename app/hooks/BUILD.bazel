
load("@npm//@bazel/typescript:index.bzl", "ts_library")
load("@npm//eslint:index.bzl", "eslint")

filegroup(
    name = "source_files",
    srcs = glob(["*.tsx"]),
)

eslint(
    name = "eslint",
    data = [
        "//:.eslintrc",
        ":source_files",
        "@npm//:node_modules",
    ],
    args = ["**/*.tsx"],
)

ts_library(
    name = "ts_library",
    srcs = [":source_files"],
    deps = [
        "@npm//@material-ui/core",
        "@npm//@types",
        "@npm//react",
    ],
    visibility = ["//visibility:public"],
)