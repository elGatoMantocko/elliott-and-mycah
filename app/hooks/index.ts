export * from './result';
export * from './scrollToTop';
export * from './serviceWorker';
export * from './theme';
