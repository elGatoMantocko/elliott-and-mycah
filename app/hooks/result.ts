import { DependencyList, useCallback, useEffect, useState } from 'react';

import { State, state } from '../models';

export enum ResultState {
  NotStarted = 'not-started',
  Pending = 'pending',
  Value = 'value',
  Error = 'error',
}

export type Result<T, U> =
  | State<ResultState.NotStarted>
  | State<ResultState.Pending>
  | State<ResultState.Value, T>
  | State<ResultState.Error, U>;

const PENDING: State<ResultState.Pending> = state(ResultState.Pending);
const NOT_STARTED: State<ResultState.NotStarted> = state(ResultState.NotStarted);

/**
 * Turns an unknown error into an {@link Error}
 * @param err unknown error
 * @returns a valid error object
 */
function handleUnknownError(err: unknown): Error {
  if (err instanceof Error) {
    return err;
  } else if (typeof err === 'string') {
    return new Error(err);
  } else if (typeof err === 'object') {
    return new Error(`encountered error in useCallableResult: ${JSON.stringify(err)}`);
  } else {
    return new Error('encountered unknown error in useCallableResult');
  }
}

/**
 * Function to get a result from an asynchronous function.
 * @param request asynchronous function to map to a result state
 * @param deps list of deps used in the request
 * @returns result state object
 */
export const useResult = <C extends () => Promise<unknown>, CValue extends Awaited<ReturnType<C>>>(
  request: C extends () => Promise<CValue> ? C : never,
  deps: DependencyList,
) => {
  const [result, setResult] = useState<Result<CValue, Error>>(NOT_STARTED);

  useEffect(() => {
    let isSubscribed = true;

    // only set state and make requests as long as this hook is still "mounted" to the dom
    // this is mostly just so we don't freak the tests out when things get unmounted
    if (isSubscribed) {
      setResult(PENDING);
      request()
        .then((data) => setResult(state(ResultState.Value, data)))
        .catch((err: unknown) => setResult(state(ResultState.Error, handleUnknownError(err))));
    }

    // if the `useResult` hook is torn down, we want to make sure we don't have any
    // lingering calls being made
    return () => {
      isSubscribed = false;
    };
    // The caller is defining the dependencies to this so we intentionally
    // want to ignore the exhaustive deps rule here.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [...deps]);
  return result;
};

/**
 * Hook to get a callable function that sets the result state. Can be reset to `not-started`
 * with the 3rd option in the tuple.
 * @param request function to call
 * @returns tuple representing the `[result-state, callable, reset]` options
 */
export const useCallableResult = <CArgs extends unknown[], CResponse>(
  request: (...args: CArgs) => Promise<CResponse>,
): [Result<CResponse, Error>, (...args: CArgs) => void, () => void] => {
  const [result, setResult] = useState<Result<CResponse, Error>>(NOT_STARTED);

  const call = useCallback(
    async (...args: CArgs) => {
      setResult(PENDING);

      try {
        const t = await request(...args);

        setResult(state(ResultState.Value, t));
      } catch (err) {
        setResult(state(ResultState.Error, handleUnknownError(err)));
      }
    },
    [request],
  );

  return [result, call, () => setResult(NOT_STARTED)];
};
