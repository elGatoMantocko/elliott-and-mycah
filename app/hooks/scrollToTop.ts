import { useEffect } from 'react';
import { useLocation } from 'react-router-dom';

/**
 * Hook that will automatically make a browser scroll to the `0` position when
 * the path changes inside of a router.
 */
export const useScrollToTop = () => {
  const { pathname } = useLocation();

  useEffect(() => {
    window.scrollTo(0, 0);
  }, [pathname]);
};
