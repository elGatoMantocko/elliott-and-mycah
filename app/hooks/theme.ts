import { createTheme, PaletteOptions, useMediaQuery } from '@mui/material';
import { useCallback, useMemo } from 'react';

export const useCustomTheme = () => {
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');

  const mode = useMemo(() => (prefersDarkMode ? 'dark' : 'light'), [prefersDarkMode]);

  const spacing = useCallback(
    (factor: number | string): number | string =>
      typeof factor === 'number' ? `${0.5 * factor}rem` : factor,
    [],
  );

  const palette = useMemo(
    (): PaletteOptions | undefined => ({
      mode,
      // TODO: play with dark mode but this site is designed for light mode
      primary: {
        dark: 'rgba(175, 126, 203, 1)',
        main: 'rgba(139, 77, 174, 1)',
        light: 'rgba(114, 78, 135, 1)',
        contrastText: '#fff',
      },
      secondary: {
        dark: 'rgba(104, 177, 104, 1)',
        main: 'rgba(31, 134, 31, 1)',
        light: 'rgba(45, 92, 45, 1)',
        contrastText: '#fff',
      },
      error: {
        dark: '#e57373',
        main: '#f44336',
        light: '#d32f2f',
        contrastText: '#000',
      },
      action: {
        // default disabled has a lot of opacity and that's stupid
        disabled: 'rgba(159, 159, 159, 1)',
        disabledBackground: 'rgba(226, 226, 226, 1)',
      },
    }),
    [mode],
  );

  return createTheme({ spacing, palette });
};
