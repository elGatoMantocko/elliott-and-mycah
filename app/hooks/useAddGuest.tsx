import { useCallableResult } from './useResult';

export const useAddGuest = () =>
  useCallableResult(
    (
      firstName: string,
      lastName: string,
      yesNo?: boolean,
      foodChoice?: 'chicken' | 'fish' | 'veg',
    ) =>
      fetch('https://sheet.best/api/sheets/4976d0ec-48c2-40ef-b4d3-740560714fb1', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ firstName, lastName, foodChoice, yesNo }),
      }),
  );
