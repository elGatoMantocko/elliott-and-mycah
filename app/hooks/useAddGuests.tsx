import { Guest } from '../models/guest';
import { useCallableResult } from './useResult';

export const useAddGuests = () =>
  useCallableResult((guests: Guest[], yesNo?: boolean) => {
    // TODO: this is a hack to test loading state
    return new Promise((resolve) => {
      setTimeout(() => resolve(JSON.stringify(guests.map((g) => ({ ...g, yesNo })))), 4000);
    });
    // fetch('https://sheet.best/api/sheets/4976d0ec-48c2-40ef-b4d3-740560714fb1', {
    //   method: 'POST',
    //   headers: {
    //     'Content-Type': 'application/json',
    //   },
    //   body: JSON.stringify(guests),
    // }),
  });
