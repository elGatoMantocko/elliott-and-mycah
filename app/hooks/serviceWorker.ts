import { useResult } from './result';

/**
 * Hook to unregister all service workers.
 * @returns true if any and all service workers were able to be unregistered
 */
export const useUnregisterServiceWorkers = () =>
  useResult(async () => {
    const registrations = await navigator.serviceWorker.getRegistrations();
    const results = await Promise.all(registrations.map((r) => r.unregister()));
    return results.length > 0 && results.every((didUnregister) => didUnregister);
  }, []);
