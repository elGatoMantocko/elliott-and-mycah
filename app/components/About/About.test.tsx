import { ThemeProvider } from '@mui/material/styles';
import { render } from '@testing-library/react';
import React, { PropsWithChildren } from 'react';
import { ParallaxProvider } from 'react-scroll-parallax';

import { useCustomTheme } from '../../hooks';
import { About } from '.';

/**
 * The About page uses a parallax banner and mui breakpoints. So we need to
 * provide a theme provider and parallax provider.
 * @param param0 props
 * @param param0.children passed to the providers
 * @returns react element
 */
const MockProviders = ({ children }: PropsWithChildren<unknown>) => {
  const theme = useCustomTheme();
  return (
    <ThemeProvider theme={theme}>
      <ParallaxProvider>{children}</ParallaxProvider>
    </ThemeProvider>
  );
};

it('should render an <About />', async () => {
  const el = render(
    <div data-testid="basic-about">
      <About />
    </div>,
    { wrapper: MockProviders },
  );

  expect(await el.findByTestId('basic-about')).toMatchSnapshot('basic-about');
});
