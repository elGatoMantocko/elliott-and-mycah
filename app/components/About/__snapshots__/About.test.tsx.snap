// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render an <About /> > basic-about 1`] = `
<div
  data-testid="basic-about"
>
  <div
    class="MuiBox-root css-1kykfr"
  >
    <div
      style="position: relative; overflow: hidden; width: 100%; height: 100%; background-color: rgb(204, 204, 204);"
    >
      <div
        style="position: absolute; top: -300px; left: 0px; right: 0px; bottom: -300px; will-change: transform; transform: translateY(300px);"
      >
        <div
          class="MuiBox-root css-1opgm91"
        >
          <img
            src="/app/images/elliottandmycah.webp"
            style="opacity: 0; visibility: hidden; grid-column: 1; grid-row: 1; display: none; width: 100%; height: 100%; object-fit: cover; object-position: center;"
          />
          <img
            src="/app/images/elliottandmycah-small.webp"
            style="grid-column: 1; grid-row: 1; opacity: 0; position: relative; width: 100%;"
          />
        </div>
      </div>
      <div
        style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; will-change: transform;"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1jyc5fq-MuiContainer-root"
        >
          <div
            class="MuiStack-root css-1jl7z82-MuiStack-root"
          >
            <h1
              class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter css-13irxxv-MuiTypography-root"
            >
              Mycah & Elliott
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1iin24e-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-r4ywzk"
    >
      <div
        class="MuiBox-root css-htt6nx"
      >
        <div
          class="MuiBox-root css-1aofbdg"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter css-1jyc7ph-MuiTypography-root"
          >
            How we met
          </h3>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter MuiTypography-gutterBottom css-1c3pd2j-MuiTypography-root"
          >
            We met through the band program while attending Purdue University. We remained
              friends until our final year when we had our first date at a farm to table
              restaurant in downtown Lafayette.
          </p>
        </div>
        <div
          class="MuiBox-root css-117vclv"
        >
          <div
            class="MuiBox-root css-1safwja"
          >
            <img
              alt="Mycah and Elliott attend a Slayter concert on the hill"
              src="/app/images/on_the_hill.jpg"
              style="max-width: 100%;"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-htt6nx"
      >
        <div
          class="MuiBox-root css-1aofbdg"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter css-1jyc7ph-MuiTypography-root"
          >
            Chicago and Seattle
          </h3>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter MuiTypography-gutterBottom css-1c3pd2j-MuiTypography-root"
          >
            We both, coincidentally, had already planned to move to Chicago, IL after graduation In
              August 2018, we moved to Seattle, WA. We lived in the South Lake Union area until we
              bought our house in Burien in September 2019. In May 2019, we added our chocolate lab,
              Ellie, to our family.
          </p>
        </div>
        <div
          class="MuiBox-root css-117vclv"
        >
          <div
            class="MuiBox-root css-1safwja"
          >
            <img
              alt="Ellie strikes a pose in front of 14000ft Mt. Rainier"
              src="/app/images/ellie_and_rainier.jpg"
              style="max-width: 100%;"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-htt6nx"
      >
        <div
          class="MuiBox-root css-1aofbdg"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter css-1jyc7ph-MuiTypography-root"
          >
            Engagement
          </h3>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter MuiTypography-gutterBottom css-1c3pd2j-MuiTypography-root"
          >
            In August 2019, we took an Alaskan cruise. While ported in Juneau, AK, we had planned
              to hike to a view of Mendenhall glacier followed by a humpback whale watching tour.
              After nearly missing the bus ride to the glacier, losing our group as soon as we
              arrived, and with the rain pouring down, we hiked a mile and half to a gravel beach
              framed by Mendenhall Glacier, a turquoise lake, and a glacier-fed waterfall. Elliott
              asked our tour guide to take a photo of us and after a few snapshots, got down on one
              knee to pop the question.
          </p>
        </div>
        <div
          class="MuiBox-root css-117vclv"
        >
          <div
            class="MuiBox-root css-1safwja"
          >
            <img
              alt="Elliott pops the question in front of the Mendenhall glacier"
              src="/app/images/one_knee.jpg"
              style="max-width: 100%;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
